File: code/09-subroutines/example_9.51-55/Events.cs
// Examples 9.51 and 9.52
// Under mono, must be compiled with "mcs Events.cs -pkg:gtk-sharp-2.0"
// On my Mac, also required the addition of several dll-s to mono's config file.
// See http://stackoverflow.com/questions/13885454/mono-on-osx-couldnt-find-gtksharpglue-2-dll

using System;
using System.Threading;
using Gtk;

// The Counter class is given a window and a label.
// It maintains an internal counter, which it increments once a second,
// and displays on the given label in the given window.
//
class Counter {
    volatile bool running = true;
    Label myLabel;
    public Counter(Window w, Label l) {
        myLabel = l;
        myLabel.Text = "0";
    }
    public void Run() {
        int n = 0;
        while(true) {
            myLabel.Text = n.ToString();
            Thread.Sleep(1000);             // milliseconds
            if (running) n++;
        }
    }
    public void Pause() {
        running = false;
    }
    public void Resume() {
        running = true;
    }
}

// Class Demo, derived from Window, contains a label and two buttons.
// The label houses a Counter (see above).
// The two buttons pause and resume the counter.
// In the code here, the pause button is created in the style of Example
// 8.59; the resume button is created in the style of Example 8.60.
//
class Demo : Window {
    Thread myThread;
    Counter myCount;

    void Paused(object sender, EventArgs a) {
        myCount.Pause();
    }

    public Demo(string title) : base(title) {
        Resize(240,120);

        // Handler for window deletion
        DeleteEvent += delegate(object sender, DeleteEventArgs a) {
            myThread.Abort();
            myThread.Join();
            Application.Quit();
            a.RetVal = true;        // not strictly required
        };

        Label myLabel = new Label();
        Button pauseButton = new Button("pause");
        pauseButton.Clicked += new EventHandler(Paused);

        Button resumeButton = new Button("resume");
        resumeButton.Clicked += delegate(object sender, EventArgs a) {
            myCount.Resume();
        };

        // Window layout
        VBox myBox = new VBox();
        myBox.PackStart(myLabel, false, false, 10);
        myBox.PackStart(pauseButton, false, false, 10);
        myBox.PackStart(resumeButton, false, false, 10);
        Add(myBox);     // Add to window

        myCount = new Counter(this, myLabel);

        // create worker thread to do the counting:
        myThread = new Thread(new ThreadStart(myCount.Run));
        myThread.Start();
    }
}

public class Events {
    public static void Main() {
        Application.Init();

        Demo myWin = new Demo("Event demo");
        myWin.ShowAll();

        Application.Run();      // become an event handler thread
    }
}

File: code/09-subroutines/example_9.51-55/Events.java
// Examples 9.53, 9.54, and 9.55

import javafx.application.Application;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;

// The Counter class is given a window and a label.
// It maintains an internal counter, which it increments once a second,
// and displays on the given label in the given window.
//
class Counter extends Thread {
    volatile Boolean running = true;
    Label myLabel;

    public Counter(Label l) {
        myLabel = l;
    }

    public void run() {
        int n = 0;
        while(true) {
            final int m = n;
            Platform.runLater(new Runnable() {
                public void run() {
                    myLabel.setText(Integer.toString(m));
                }
            });
            try {
                sleep(1000);   // milliseconds
            } catch(InterruptedException e) {};
            if (running) n++;
        }
    }

    public void pauseMe() {
        running = false;
    }

    public void resumeMe() {
        running = true;
    }
}

// Class Events, derived from Application, contains a label and two
// buttons.  The label houses a Counter (see above).
// The two buttons pause and resume the counter.
// In the code here, the pause button is created in the style of Example
// 8.61; the resume button is created in the style of Example 8.62.
//
public class Events extends Application {
    public static void main(String[] args) {
        launch(args);
    }

    Counter countThread;

    class PauseListener implements EventHandler<ActionEvent> {
        public void handle(ActionEvent e) {
            countThread.pauseMe();
        }
    }

    public void start(Stage stg) {
        stg.setTitle("Event demo");
        stg.setOnCloseRequest(e -> {
            System.exit(0);
        });

        Label countLabel = new Label("0");

        Button pauseButton = new Button();
        pauseButton.setText("pause");
        pauseButton.setOnAction(new PauseListener());

        Button resumeButton = new Button();
        resumeButton.setText("resume");
        // The following also works; commented out in favor of the
        // subsequent lambda expression-based version.
        /*
        resumeButton.setOnAction(new EventHandler<ActionEvent>() {
            public void handle(ActionEvent e) {
                countThread.resumeMe();
            }
        });
        */
        resumeButton.setOnAction(e -> {
            countThread.resumeMe();
        });

        VBox root = new VBox();
        root.getChildren().add(pauseButton);
        root.getChildren().add(resumeButton);
        root.getChildren().add(countLabel);
        stg.setScene(new Scene(root, 300, 300));
        stg.show();

        // create worker thread to do the counting:
        countThread = new Counter(countLabel);
        countThread.start();
    }
}
